src/Security/TwoFactor/ContaoTwoFactorAuthenticator.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Security\TwoFactor;
+
+use BaconQrCode\Renderer\Image\SvgImageBackEnd;
+use BaconQrCode\Renderer\ImageRenderer;
+use BaconQrCode\Renderer\RendererStyle\RendererStyle;
+use BaconQrCode\Writer;
+use Contao\User;
+use OTPHP\TOTP;
+use ParagonIE\ConstantTime\Base32;
+use Symfony\Component\HttpFoundation\Request;
+
+class ContaoTwoFactorAuthenticator implements ContaoTwoFactorAuthenticatorInterface
+{
+    /**
+     * @param User   $user
+     * @param string $code
+     *
+     * @return bool
+     */
+    public function validateCode(User $user, string $code): bool
+    {
+        // The 2FA app from Google (Google authenticator) does not strictly confirm to RFC 4648 [1] (they confirm to the old RFC 3548 [2]).
```

Comment:
```php
private function getUpperUnpaddedScretForUser(User $user): string
```

Then you need to add the google comment only once :-P

Reply:
True that.